dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -663,6 +663,322 @@ template class NaryOpNode<functional::min<double>>;
 template class NaryOpNode<std::multiplies<double>>;
 template class NaryOpNode<std::plus<double>>;
 
+// PartialReduceNode *****************************************************************
+std::vector<ssize_t> partial_reduce_shape(const std::span<const ssize_t> input_shape,
+                                          const ssize_t axis) {
+    std::vector<ssize_t> shape;
+    shape.assign(input_shape.begin(), input_shape.end());
+    shape.erase(shape.begin() + axis);
+    return shape;
+}
+
+std::vector<ssize_t> as_contiguous_strides(const std::span<const ssize_t> shape) {
+    ssize_t ndim = static_cast<ssize_t>(shape.size());
+
+    assert(ndim >= 0);
```

Comment:
```c++
    assert(ndim > 0);
```
right? because we access `strides[ndim - 1]` shortly.

Or, if we want to handle the `ndim == 0` case, we could do
```c++
    assert(ndim >= 0);
    if (ndim < 1) return {};
```